packages/core/src/browser/tree/tree-iterator.ts

```diff
@@ -5,71 +5,145 @@
  * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
  */
 
-import { ITreeNode, ICompositeTreeNode } from "./tree";
-import { IExpandableTreeNode } from "./tree-expansion";
+import { TreeNode, CompositeTreeNode } from "./tree";
+import { ExpandableTreeNode } from "./tree-expansion";
 
-export interface ITreeNodeIterator extends Iterator<ITreeNode | undefined> {
+export interface TreeIterator extends Iterator<TreeNode> {
 }
 
-export namespace ITreeNodeIterator {
-    export interface IOptions {
+export namespace TreeIterator {
+
+    export interface Options {
         readonly pruneCollapsed: boolean
     }
-    export const DEFAULT_OPTIONS: IOptions = {
+
+    export const DEFAULT_OPTIONS: Options = {
         pruneCollapsed: false
     };
+
 }
 
-export abstract class AbstractTreeNodeIterator implements ITreeNodeIterator {
-    constructor(protected node: ITreeNode | undefined,
-        protected readonly options = ITreeNodeIterator.DEFAULT_OPTIONS) {
+export abstract class AbstractTreeIterator implements TreeIterator, Iterable<TreeNode> {
+
+    protected readonly delegate: IterableIterator<TreeNode>;
+
+    constructor(protected readonly root: TreeNode, protected readonly options: TreeIterator.Options = TreeIterator.DEFAULT_OPTIONS) {
+        this.delegate = this.iterator(this.root);
     }
 
-    next(): IteratorResult<ITreeNode | undefined> {
-        if (!this.node) {
-            return {
-                value: undefined,
-                done: true,
-            };
-        }
-        this.node = this.doNext(this.node);
-        return {
-            value: this.node,
-            done: false
-        };
+    [Symbol.iterator]() {
+        return this.delegate;
     }
 
-    protected abstract doNext(node: ITreeNode): ITreeNode | undefined;
+    next(): IteratorResult<TreeNode> {
+        return this.delegate.next();
+    }
 
-    protected hasChildren(node: ITreeNode | undefined): node is ICompositeTreeNode {
-        if (!ICompositeTreeNode.is(node)) {
-            return false;
-        }
-        if (node.children.length === 0) {
-            return false;
+    protected abstract iterator(node: TreeNode): IterableIterator<TreeNode>;
+
+    protected children(node: TreeNode): TreeNode[] | undefined {
+        if (!CompositeTreeNode.is(node)) {
+            return undefined;
         }
-        if (this.options.pruneCollapsed) {
-            return IExpandableTreeNode.isExpanded(node);
+        if (this.options.pruneCollapsed && this.isCollapsed(node)) {
+            return undefined;
         }
-        return true;
+        return node.children.slice();
+    }
+
+    protected isCollapsed(node: TreeNode): boolean {
+        return ExpandableTreeNode.isCollapsed(node);
+    }
+
+    protected isEmpty(nodes: TreeNode[] | undefined): boolean {
+        return nodes === undefined || nodes.length === 0;
+    }
+
+}
+
+export class PreOrderTreeIterator extends AbstractTreeIterator {
```

Comment:
just suggestion: `DepthFirstTreeIterator`? We could add options later to configure an order mode later.